### DISCUSSION TITLE
Zero penetration depth with two equal boxes on same position

### DISCUSSION THREAD

**Discussion by maxwave11** [coal-library/coal#578] [Q&A] [answered]
Hi everyone! In code below I'm trying to test collision between two equal boxes on same position (without transformation) and **penetration depth is zero**. 

```C++
shared_ptr<fcl::Box> shape1 = make_shared<fcl::Box>(1,1,1);
shared_ptr<fcl::Box> shape2 = make_shared<fcl::Box>(1,1,1);

auto* obj1 = new fcl::CollisionObject(shape1);
auto* obj2 = new fcl::CollisionObject(shape2);

fcl::CollisionRequest col_req;
fcl::CollisionResult col_res;
    
// Collision call
fcl::collide(obj1, obj2, col_req, col_res);
    
cout << "Collision? " << col_res.isCollision() << "\n";
if (col_res.isCollision()) {
    fcl::Contact contact = col_res.getContact(0);
    cout << "Penetration depth: " << contact.penetration_depth << "\n";
    cout << "Witness point on shape1: " << contact.pos << "\n";
    cout << "Normal: " << contact.normal << "\n";
}
```

Output:

_Collision? 1
Penetration depth: 0
Witness point on shape1:   0  0 0.5
Normal: 0 0 1_

What I'm doing wrong? 

Although I tested the same case with old FCL library and I'm getting non-zero value.
I'm trying to use hpp-fcl for clash detection between two arbitrary objects (including concave).
Could you please explain what exact means penetration depth and how can I visualize it? I need to decide is this value fits for clash detection, when object A can't penetrate to B on some depth (inner margin) limit.

**Comment by lmontaut** [comment, marked as answer]
Hi @maxwave11,
We are working on the next major release of hpp-fcl. If I try your code with the latest commit of the `devel` branch, I get a penetration depth of `-1.0` like expected.
Also, quick note: `contact.pos` is not the witness point on shape1.
We have `contact.pos = (p1 + p2) / 2` where `p1 = contact.nearest_points[0]` and `p2 = contact.nearest_points[1]`, hence `p1` and `p2` are the witness points on the shapes.

**Reply by maxwave11** [reply]
It helps thank you for quick answer!
